### DISCUSSION TITLE
Access keyboard and mouse state like in WPF

### DISCUSSION THREAD

**Discussion by mykolamor** [AvaloniaUI/Avalonia#5487] [Q&A]
Hi! I'm currently trying to translate code from WPF to Avalonia and I ran into problems with WPF Keyboard and Mouse classes.
I'm looking for a replacement for `Keyboard.Modifiers` and `Mouse.LeftButton` etc. for methods that are not direct mouse and keyboard event handlers. For example, [this line for keyboard](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexEditor.xaml.cs#L4342) and [this line for mouse](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexEditor.xaml.cs#L4280).

P.S. As a follow-up to #5470, is there a way to access system properties like in WPF:
1. `VerticalScrollBar.Track` [here](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexEditor.xaml.cs#L3964)
2. `SystemParameters.WheelScrollLines` [here](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexEditor.xaml.cs#L4360)
3. `Visual.TransformToAncestor` [here](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexByte.cs#L194)
Maybe `Point point = TransformToAncestor(_parent).Transform(new Point(0, 0))` is identical to `Point point = new Point(0, 0).Transform(this.TransformToVisual(_parent))`?
4. Is WPF `FrameworkElement.LayoutTransform` the same as Avalonia `Visual.RenderTransform` as [here](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexEditor.xaml.cs#L5349)?

**Comment by jp2masa** [comment]
1. I think there's no concept of "track". The `Value`, `Maximum` and `Minimum` are properties of the `ScrollBar` (derived from `RangeBase`).
2. AFAIK that's not currently possible.
3. I think that's equivalent to `Matrix m = Matrix.CreateTranslation(0, 0) * this.TransformToVisual(_parent).Value;` and then creating a point: `Point p = new Point(m.M31, m.M32);`, but check if it works first (for example with points `(1, 0), (0, 1), (2, 1)` and compare with WPF).
4. Use [`LayoutTransformControl`](https://github.com/AvaloniaUI/Avalonia/blob/cae9badc1c3220baa4400c3db3a3e548075c50fe/src/Avalonia.Controls/LayoutTransformControl.cs).

**Reply by mykolamor** [reply]
@jp2masa what about keyboard and mouse?

**Reply by jp2masa** [reply]
That's all available from `PointerEventArgs`, for example:

```cs
e.KeyModifiers; // keyboard
e.GetCurrentPoint(this).Properties.IsLeftButtonPressed; // mouse
```

**Reply by mykolamor** [reply]
> 
> 
> That's all available from `PointerEventArgs`, for example:
> 
> ```cs
> e.KeyModifiers; // keyboard
> e.GetCurrentPoint(this).Properties.IsLeftButtonPressed; // mouse
> ```

But the thing is, the code at WPF gets access to keyboard and mouse without having PointerEventArgs available. For example, [here](https://github.com/abbaye/WpfHexEditorControl/blob/9af1d1dacd4cd5528de7e83024a4aa77fc9e3af0/Sources/WPFHexaEditor/HexEditor.xaml.cs#L4280).

**Reply by maxkatz6** [reply]
> code at WPF gets access to keyboard and mouse without having PointerEventArgs available

Which is not a good practice.
Also you are free to store and reuse last values of those events.

**Reply by mykolamor** [reply]
So is my best bet right now with #4447, #4062 and device state checks is to write my own InputDevice (since KeyboardDevice will be removed before 1.0) based on the existing one add the required functionality there?

**Reply by workgroupengineering** [reply]
> Which is not a good practice.
> Also you are free to store and reuse last values of those events.

Why?

I don't understand why to dirty the ViewModel to intercept events to memorize the status of some keys.

I prefer in such a thing:

```csharp
class MyVM:BaseViewModel
{
    readonly IInputManager _currentInputManager;
    public MyVm(IInputManager manager) => 
       _currentInputManager = manager;
       
    public void Do(object parameter)
    {
         var capslockState = _currentInputManager
                .PrimaryKeyboardDevice.GetKeyStates(VirtualKey.CapsLock)
         switch (capslockState)
         {
           .....
         }
    }
} 
```

**Reply by maxkatz6** [reply]
> PrimaryKeyboardDevice

My concern was with multiple devices (and sometimes multiple types of them).
In some cases defining and using "primary" device will solve this concern. Not saying that it's unusual to have two keyboards (there is on screen virtual keyboards though).

But with pointer input devices it's more complicated since user can have mouse and stylus/pen, both can be used in some drawing application and using "primary" one will be questionable.

As you have InputManager with primary keyboard device, it could be possible to have input manager with all connected devices with ability to read their state. Although it probably won't work well on some specific OSs (I don't have information about this at least).

**Reply by kekekeks** [reply]
I'd suggest to have a separate "NonPortableApis" package that would expose whatever APIs it wants

**Reply by AngryCarrot789** [reply]
If avalonia can differentiate physical keyboards and mice and other input devices, there should be lists of devices it has received events from, including by default the input devices connected at application startup. This way, we can access the first keyboard(s) or mouse the operating system has. And since most computers only have 1 keyboard and mouse, those lists would have 1 element at startup.

E.g. `MouseDevice.Devices[0].IsLeftButtonPressed`

This issue has been a big road block, because the behaviour of one of my controls needs to update a copy of the mouse position when they user clicks the escape key
